timemachine/fe/atom_mapping.py

```diff
@@ -236,6 +243,79 @@ def _deduplicate_all_cores(all_cores):
 
 
+def ring_forming_breaking_count(mol_a, mol_b, core: NDArray) -> int:
+    """Counts the number of rings broken and formed given a core. Does not count ring insertion/deletions. Considers
+    the number number of rings broken/formed in both endstates, reversing the mol order and the core order produces
+    the same value.
+
+    The algorithm is to add all ring atoms/bonds to a graph, and count the cycles that contain both core and mol a or
+    mol b atoms.
+    """
+    g = nx.Graph()
+    core_atoms_a = set(core[:, 0])
+    core_atoms_b = set(core[:, 1])
+    for atom in mol_a.GetAtoms():
+        idx = atom.GetIdx()
+        if atom.IsInRing():
+            if idx in core_atoms_a:
+                g.add_node(idx, atom_type=AtomMapFlags.CORE)
+            else:
+                g.add_node(idx, atom_type=AtomMapFlags.MOL_A)
+    for atom in mol_b.GetAtoms():
+        idx = atom.GetIdx()
+        if atom.IsInRing():
+            if idx in core_atoms_b:
+                g.add_node(idx + mol_a.GetNumAtoms(), atom_type=AtomMapFlags.CORE)
+            else:
+                g.add_node(idx + mol_a.GetNumAtoms(), atom_type=AtomMapFlags.MOL_B)
+
+    # if there are no ring atoms, return 0
+    if len(g.nodes) == 0:
+        return 0
+
+    for bond in mol_a.GetBonds():
+        if bond.IsInRing():
+            src, dst = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+            if src in g.nodes and dst in g.nodes:
+                g.add_edge(src, dst)
+
+    for bond in mol_b.GetBonds():
+        if bond.IsInRing():
+            src, dst = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+            src += mol_a.GetNumAtoms()
+            dst += mol_a.GetNumAtoms()
+            if src in g.nodes and dst in g.nodes:
+                g.add_edge(src, dst)
+
+    # Need at least three edges to make a cycle
+    if len(g.edges) < 3:
+        return 0
+
+    core_breaks = 0
+    for i, comp in enumerate(nx.connected_components(g)):
+        subgraph = g.subgraph(comp)
+        if all([data["atom_type"] == AtomMapFlags.CORE for _, data in subgraph.nodes(data=True)]):
+            continue
+
+        for atom_type in [AtomMapFlags.MOL_A, AtomMapFlags.MOL_B]:
+            ring_atoms = subgraph.subgraph(
+                [n for n, data in subgraph.nodes(data=True) if data["atom_type"] != atom_type]
+            )
+            ring_atom_is_core = [data["atom_type"] == AtomMapFlags.CORE for _, data in ring_atoms.nodes(data=True)]
+            # If no atoms are core atoms, it is not a ring forming/breaking transformation and just an insertion
+            if not any(ring_atom_is_core):
+                continue
+            # If the ring atoms are all core atoms, no ring break has occurred
+            if all(ring_atom_is_core):
+                continue
+            try:
+                nx.find_cycle(ring_atoms)
+                core_breaks += 1
+            except nx.exception.NetworkXNoCycle:
+                pass
+    return core_breaks
```

Comment:
nit: name seems misleading, IIUC: is this counting the number of breaks and formations, and not just in the core?

Reply:
Renamed in https://github.com/proteneer/timemachine/pull/1363/commits/8728b64ecc6fb30ff546bd1209016e31afd0091d